seastar/src/ffi_utils.rs

```diff
@@ -15,6 +15,28 @@ where
     fn_once_caller::<Func, Ret>
 }
 
+/// Convert the pointer to Box<Func> and then call it but don't consume.
+fn fn_caller<Func, Ret>(raw_func: *mut u8) -> Ret
+where
+    Func: Fn() -> Ret,
+{
+    unsafe {
+        let boxed_func = Box::from_raw(raw_func as *mut Func);
+        let ret = boxed_func();
+        std::ptr::swap(raw_func, Box::into_raw(boxed_func) as *mut u8);
```

Comment:
What is this line supposed to do?

https://doc.rust-lang.org/std/ptr/fn.swap.html

This line just swaps two bytes.

Reply:
Well, putting the function in a `Box` meant that the `Box` now made sure to drop it, but we didn't want that so that `raw_func` would still point to valid data. So I extracted the data out of the `Box` and wanted to put it back, which I opted to do with `swap`

Reply:
This won't be needed as I followed your advice in the other thread, but tell me if my reasoning was flawed.

Reply:
`Box::from_raw` creates a Box which assumes ownership of the pointer you give to it. That Box will not move memory, just use the same pointer.

`Box::into_raw` just forgets about the `Box` wrapper and returns the pointer it contains. It does not "extract" any data behind the pointer.

`std::ptr::swap` takes two pointers and swaps the memory that they point to, based on their type. Given that they are of type `*mut u8` here, the function will just take the bytes pointed by them and swap them together. More specifically, `raw_func` and `Box::into_raw(boxed_func) as *mut u8` should be equal so the function will not do anything useful.

It would make more sense to call `std::ptr::swap` on two pointers of type `*mut Func`. Currently, you risk triggering undefined behavior if `Func` is a ZST (zero-sized type).

Besides, I don't think you should really touch the memory behind `raw_func`. `raw_func` should be a const pointer, as the function behind `raw_func` implements `Fn()` and possibly `Sync`, so multiple threads/shards might use it at the same time.

Reply:
I see, so I could have just done `let _ = Box::from_raw(boxed_func)`.
I changed `raw_func` to be a pointer to `const`, but this required me to write a sort of doubtfully pretty `get_dropper_const`. Take a look and see if we're content with that.